### PR TITLE
Add Reddit Scraper

### PR DESCRIPTION
**Pull request by Max-Leopold**
Purpose: Generate resources by scraping the reddit frontpage. Post these resources on a Kafka topic.

Posts: The Reddit-Scraper scrapes the frontpage of old.reddit.com (r/all). A post is a reddit post. title is the title of the post, author the u/... user and the date is the date of the creation. Every post is his own resource.

Comment: The scraper clicks on every reddit post to see the comments under the post. Every comment is his own resource.

### CODE REVIEW

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/kafka/Producer.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static final Logger logger = LoggerFactory.getLogger(Producer.class);
+
+    @Autowired
+    private KafkaTemplate<String, Resource> kafkaTemplate;
+
+    public void send(Resource resource){
+
+        logger.info("sending message='{}'", resource.toString());
+        kafkaTemplate.send("reddit", String.valueOf(System.currentTimeMillis()), resource);
```
Comment (YNedderhoff) [inline reddit-scraper/src/main/java/de/brandwatch/redditscraper/kafka/Producer.java]: system time again

**Comment by Max-Leopold** [comment]
Did that. If you need a more detailed explanation I can also do that.

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForComments.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    public RedditScraperForComments(Producer producer) {
+        this.producer = producer;
+    }
+
+    public void scrapeRedditForComments(String URL) throws IOException {
+
+        Document reddit = Jsoup.connect(URL).get();
+
+        Elements elements = reddit.getElementsByClass("entry unvoted");
```
[the remark this answers is not in the document]
Comment (Max-Leopold) [reply reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForComments.java]: just a reddit website class. Don't know why it's named that way.

**Comment by Max-Leopold** [comment]
All done.

File: mentions-generator/src/main/java/de/brandwatch/minianalytics/mentiongenerator/kafka/Consumer.java [hunk 3]
```diff
@@ -18,7 +18,7 @@ public Consumer(LuceneService luceneService) {
         this.luceneService = luceneService;
     }
 
-    @KafkaListener(topics = "resources")
+    @KafkaListener(topics = {"twitter, reddit"})
```
Comment (YNedderhoff) [inline mentions-generator/src/main/java/de/brandwatch/minianalytics/mentiongenerator/kafka/Consumer.java]: You wanted to revert this

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                .map(this::mapToResource)
+                .filter(resource -> !(resource.getDate() == null))
+                .forEach(producer::send);
+
+        Elements nextPage = reddit.select(".next-button > a");
+
+        for (Element element : nextPage) {
+            pageCounter = pageCounter + 1;
+            //Scrape top 50 sites
+            if(pageCounter < 51) scrapeRedditForTitles(element.attr("href"), pageCounter);
```
Comment (YNedderhoff) [inline reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: This works, but we normally do 

```
if () {

}
```

style. If you don't want to you don't have to change it here, but you might as well get used to it.

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Scheduled(fixedDelay = 1000 * 30)
+    public void scrapeReddit() throws IOException {
+        scrapeRedditForTitles(REDDIT_URL, 1);
+    }
+
+    public void scrapeRedditForTitles(String url, int pageCounter) throws IOException {
+        Document reddit = Jsoup.connect(url).get();
+        Elements posts = reddit.getElementsByClass("top-matter"); //Whitespace at the end is important!
```
Comment (YNedderhoff) [inline reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: Doesn't seem to be used

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        this.producer = producer;
+        this.redditScraperForComments = redditScraperForComments;
+    }
+
+    @Scheduled(fixedDelay = 1000 * 30)
+    public void scrapeReddit() throws IOException {
+        scrapeRedditForTitles(REDDIT_URL, 1);
+    }
+
+    public void scrapeRedditForTitles(String url, int pageCounter) throws IOException {
```
Comment (YNedderhoff) [inline reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: does this need to be public?

File: mentions-generator/src/main/java/de/brandwatch/minianalytics/mentiongenerator/kafka/Consumer.java (hunk 3 above)
Comment (Max-Leopold) [reply mentions-generator/src/main/java/de/brandwatch/minianalytics/mentiongenerator/kafka/Consumer.java]: Done

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java (hunk 4 above)
Comment (Max-Leopold) [reply reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: Changed it

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java (hunk 5 above)
Comment (Max-Leopold) [reply reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: Removed

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java (hunk 6 above)
Comment (Max-Leopold) [reply reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: No

File: reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+public class RedditScraperForTitles {
+
+    private static final Logger logger = LoggerFactory.getLogger(RedditScraperForTitles.class);
+
+    private static final String REDDIT_URL = "https://old.reddit.com";
+
+    private Set<String> visitedSites = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (Max-Leopold) [reply reddit-scraper/src/main/java/de/brandwatch/redditscraper/reddit/RedditScraperForTitles.java]: Removed

**Pull request merged**